appium-tests/helpers/contactsHelper.js

```diff
@@ -0,0 +1,222 @@
+
+    var result =
+        'var results = document.getElementById("info");\n' +
+        'var obj = new ContactFindOptions();\n' +
+        'obj.filter = ' + preparedOldDisplayName + ';\n' +
+        'obj.multiple = false;\n' +
+
+        'navigator.contacts.find(["displayName", "name"], function(contacts) {\n' +
+        '    if (contacts.length === 0) {\n' +
+        '        results.innerHTML = "No contacts to update.";\n' +
+        '        return;\n' +
+        '    }\n' +
+        '    var contact = contacts[0];\n' +
+        '    contact.displayName = ' + preparedNewDisplayName + ';\n' +
+        '    contact.name = ' + preparedNewName + ';\n' +
+        '    contact.save(function(updated) {\n' +
+        '        results.innerHTML = "Contact updated.";\n' +
+        '    },function(e) {\n' +
+        '        results.innerHTML = "Update failed: error " + e.code;\n' +
+        '    });\n' +
+        '}, function(e) {\n' +
+        '    results.innerHTML = "Search failed: error " + e.code;\n' +
+        '}, obj);';
+
+    return result;
+};
+
+module.exports.getRemoveTestContactsCode = function () {
+    var result =
+        'var results = document.getElementById("info");\n' +
+        'results.innerHTML = "";\n' +
+        'var obj = new ContactFindOptions();\n' +
+        'obj.filter = "DeleteMe";\n' +
+        'obj.multiple = true;\n' +
+        'navigator.contacts.find(["note"], function(contacts) {\n' +
+        '    var removes = [];\n' +
+        '    contacts.forEach(function(contact) {\n' +
+        '        removes.push(contact);\n' +
+        '    });\n' +
+        '    if (removes.length === 0) {\n' +
+        '        results.innerHTML = "No contacts to remove";\n' +
+        '        return;\n' +
+        '    }\n' +
+
+        '   var nextToRemove;\n' +
+        '   if (removes.length > 0) {\n' +
+        '        nextToRemove = removes.shift();\n' +
+        '    }\n' +
+
+        '    function removeNext(item) {\n' +
+        '        if (typeof item === "undefined") {\n' +
+        '            return;\n' +
+        '        }\n' +
+
+        '        if (removes.length > 0) {\n' +
+        '            nextToRemove = removes.shift();\n' +
+        '        } else {\n' +
+        '            nextToRemove = undefined;\n' +
+        '        }\n' +
+
+        '        item.remove(function removeSucceeded() {\n' +
+        '            results.innerHTML += "Removed a contact with ID " + item.id + "<br/>";\n' +
+        '            removeNext(nextToRemove);\n' +
+        '        }, function removeFailed() {\n' +
+        '            results.innerHTML += "Failed to remove a contact with ID " + item.id + "<br/>";\n' +
+        '            removeNext(nextToRemove);\n' +
+        '        });\n' +
+        '    }\n' +
+        '    removeNext(nextToRemove);\n' +
+        '}, function(e) {\n' +
+        '   results.innerHTML = "Search failed: error " + e.code;\n' +
+        '}, obj);';
+
+    return result;
+};
```

Comment:
Currently the tests already depend on the Mobilespec app. This code should live in the app.

There are many reasons why writing code in strings is a bad pattern. Here are just a few:
- it's very hard to de-duplicate and refactor
- it's not checked for syntax, therefore:
    - it's easier to make mistakes in it
    - it's more difficult to fix bugs in it
    - even if it looks fine, it can break at runtime because errors can be inserted at runtime
- it's difficult to edit and read because:
    - it's not highlighted
    - its format is hard to see

In short, it's a very dangerous pattern and we can easily avoid it, so we should definitely not be using it. I agree with you that our UI code lives too far from our UI tests, but I think that's a problem with Mobilespec's design. There are many ways to fix it (e.g. having separate test apps that live with each plugin), and we should discuss fixing Mobilespec as a separate task.